kube-controllers/pkg/controllers/loadbalancer/loadbalancer_controller.go

```diff
@@ -0,0 +1,952 @@
+		svc.Annotations[annotationIPv6Pools] != "" ||
+		svc.Annotations[annotationLoadBalancerIP] != "" {
+		return true
+	}
+	return false
+}
+
+// validateAnnotation checks if the ips specified in the calico annotation are valid.
+// Each service can have at most one ipv4 and one ipv6 address
+func (c *loadBalancerController) parseAnnotations(annotations map[string]string) ([]cnet.IP, []cnet.IPNet, []cnet.IPNet, error) {
+	var loadBalancerIPs []cnet.IP
+	var ipv4Pools []cnet.IPNet
+	var ipv6Pools []cnet.IPNet
+
+	for key, annotation := range annotations {
+		switch key {
+		case annotationIPv4Pools:
+			poolCIDRs := []string{}
+			err := json.Unmarshal([]byte(annotation), &poolCIDRs)
+			if err != nil {
+				return nil, nil, nil, err
+			}
+			ipv4Pools, err = c.resolvePools(poolCIDRs, true)
+			if err != nil {
+				return nil, nil, nil, err
+			}
+			if ipv4Pools == nil {
+				return nil, nil, nil, fmt.Errorf("failed to resolve pools for IPv4 addresses from annotation")
+			}
+		case annotationIPv6Pools:
+			poolCIDRs := []string{}
+			err := json.Unmarshal([]byte(annotation), &poolCIDRs)
+			if err != nil {
+				return nil, nil, nil, err
+			}
+			ipv6Pools, err = c.resolvePools(poolCIDRs, false)
+			if err != nil {
+				return nil, nil, nil, err
+			}
+			if ipv6Pools == nil {
+				return nil, nil, nil, fmt.Errorf("failed to resolve pools for IPv6 addresses from annotation")
+			}
+		case annotationLoadBalancerIP:
+			ipAddrs := []string{}
+			err := json.Unmarshal([]byte(annotation), &ipAddrs)
+			if err != nil {
+				return nil, nil, nil, err
+			}
+			ipv4 := 0
+			ipv6 := 0
+			for _, ipAddr := range ipAddrs {
+				curr := cnet.ParseIP(ipAddr)
+				if curr == nil {
+					return nil, nil, nil, fmt.Errorf("Could not parse %s as a valid IP address", ipAddr)
+				}
+				if curr.To4() != nil {
+					ipv4++
+				} else if curr.To16() != nil {
+					ipv6++
+				}
+				loadBalancerIPs = append(loadBalancerIPs, *curr)
+			}
+
+			if ipv6 > 1 || ipv4 > 1 {
+				return nil, nil, nil, fmt.Errorf("At max only one ipv4 and one ipv6 address can be specified. Recieved %d ipv4 and %d ipv6 addresses", ipv4, ipv6)
+			}
+		}
+	}
+	return loadBalancerIPs, ipv4Pools, ipv6Pools, nil
+}
+
+// createHandle Returns a handle to use for IP allocation for the service
+func createHandle(svc *v1.Service) (string, error) {
+	prefix := "lb-"
+	handle := strings.ToLower(fmt.Sprintf("%s-%s-%s", svc.Name, svc.Namespace, svc.UID))
+
+	hasher := sha256.New()
+	_, err := hasher.Write([]byte(handle))
+	if err != nil {
+		log.WithError(err).Panic("Failed to generate hash from handle")
+		return "", err
+	}
+
+	hash := base64.RawURLEncoding.EncodeToString(hasher.Sum(nil))
+	regex := regexp.MustCompile("([-_.])")
+	hash = regex.ReplaceAllString(hash, "")
+	handle = prefix + hash
+	if len(handle) > k8svalidation.DNS1123LabelMaxLength {
+		handle = handle[:k8svalidation.DNS1123LabelMaxLength]
+	}
+
+	return handle, nil
+}
+
+func isIpInIppool(ipAddr string, cidrs []cnet.IPNet) bool {
```

Comment:
```suggestion
func isIPInIPPools(ipAddr string, cidrs []cnet.IPNet) bool {
```

(acrynyms same-case, plural since multiple pools)

Or even just `poolsContain(cidrs, ipAddr)`